### ISSUE TITLE
Allow to set custom omnifunc or completefunc in config and describe it.

### ISSUE THREAD

**Issue by pawelurbanski** [echasnovski/mini.nvim#586]
### Contributing guidelines

- [X] I have read [CONTRIBUTING.md](https://github.com/echasnovski/mini.nvim/blob/main/CONTRIBUTING.md)
- [X] I have read [CODE_OF_CONDUCT.md](https://github.com/echasnovski/mini.nvim/blob/main/CODE_OF_CONDUCT.md)

### Module(s)

Mini.completion

### Description

Dear Evgeni,
I've setup Mini.completion and it is working fantasticly well with hte LSP integration. Since I have the need and work on a completion provider for a custom format, I wanted to go down the custom omnifunc or completefunc path.
After studying the sources I see the following potential options, which I would like to confirm with you:
1. Use /after/ftplugin/<format>.lua file and setup Mini.completion there,
** Disable automatic setup, which is an autocommand that sets up the default omnifunc,
** Assign, for example: fource_func = 'v:lua.customFormatOmnifunc' in the config values,
2. Create an autocommand for my filetype that will setup my custom omnifunc as it is done in the plugin.
** The omnifunc setup function uses vim.bo[config.lsp_completion.source_func] key - should I use it verbatim?
** Will returning just the regular completion items be enough or I should get a handle to the plugin to copy the hints with results and sent somehow?

All the best from Warsaw,
Pawel

**Comment by echasnovski** [comment]
Hello, Pawel!

Would you mind clarifying a bit on what functionality you want to have in the end? Like, in the form of "Here is an example of text. I would like to press this key and see this result.".

The 'mini.completion' module will only trigger lsp completion [if source function is set to the one from 'mini.completion'](https://github.com/echasnovski/mini.nvim/blob/baf17e6472b0968661dce5daa372251e934da39f/lua/mini/completion.lua#L695-L699). So simply setting custom 'omnifunc' or 'sourcefunc' will not be enough. It is not really designed for anything else outside its two-stage setup.

If you really want to have some custom completion being autotriggered only in certain files, then I would suggest trying to go the route 1 with 'after/ftplugin/xxx.lua' file. Inside it, I'd create `vim.b.minicompletion_config` variable with custom `fallback_action`. It can be either an actual function that displays completion popup or string of keys to emulate. I am afraid without more details I can not help more here.

**Comment by pawelurbanski** [comment]
Evgeni,
I would like to swap your lsp omnifunc used by default for the following case:
1. Use the Mini.completion mechanism of automatically triggering completion on keystrokes,
2. Displaying the results in the native menu,
3. Provide my own omnifunc function that will:
** Read the current line and parse it,
** Get the context and find completion results,
** Return them in the Neovim expected format.

I tried the following:
```lua
require(mini.completion').setup({
  lsp_completion = {
    source_func = 'omnifunc',
    auto_setup = false
  },
})
-- Use custom omnifunc.
vim.bo.omnifunc = 'v:lua.bufferOmnifunc'
```

The `bufferOmnifunc` is very simple. It is an Neovim global function with 2 parameters just like the default omnifunc that writes to a log file just to check if it i sbeing used.

Unfortunately it looks like it is not being picked. The require call is in the `ftplugin/xxxx.lua` file.

Thank you for your time and effort...

**Comment by echasnovski** [comment]
> 3\. Provide my own omnifunc function that will:

Yeah, sorry, but plugging custom omnifunction instead of the one from 'mini.completion' wouldn't work for reasons I described earlier ('mini.completion' checks for source function to be appropriately set to one from this module).

My best suggestion here would be to sacrifice the fallback and essentially use one-step completion with your omnifunc as active. Here is an example of what to put inside 'after/ftplugin/xxxx.lua':
```lua
-- Define buffer-local config.
-- This will skip LSP step and fallback on calling 'omnifunc'.
vim.b.minicompletion_config = {
  lsp_completion = { auto_setup = false },
  fallback_action = '<C-x><C-o>',
}

-- Setup omnifunc. This should be a function reachable from
-- global namespace `_G` (not necessarily directly in it),
-- because it is the only reasonable way to set it as omnifunc.
_G.my_filetype_omnifunc = function()
  -- Your code for omnifunc goes here. Ideally it should result into showing
  -- built-in completion popup menu.
  -- Beware that 'mini.completion' is not designed to handle this use case.
end
vim.bo.omnifunc = 'v:lua._G.my_filetype_omnifunc'
```

**Comment by pawelurbanski** [comment]
Dear Evgeni,
I will try to create a custom omnifunc or completefunc and register it per my filetype. Since I am not that advanced in Neovim some things are not as obvious to me.
I asked about my case for I assumed that once I stick to the omnifunc or completefunc signiture and return type format Mini.completion will be able to function like a completion popup manager.

I will try to take small steps practicing with a plaintext file.
When browsing through code I found the following buffer variable: `vim.b.minicompletion_disable`.

Will it disable the mechanism or I need to apply the config you suggested to disable automatic setup. We would need to handle the fallback case as well for it uses the standard keys.

All the best,
Pawel

**Comment by echasnovski** [comment]
> I asked about my case for I assumed that once I stick to the omnifunc or completefunc signiture and return type format Mini.completion will be able to function like a completion popup manager.

Sorry, not quite. The best approach I can think of is the one from previous comment.



> Will it disable the mechanism or I need to apply the config you suggested to disable automatic setup. We would need to handle the fallback case as well for it uses the standard keys.

Setting this variable to `true` will completely disable all 'mini.completion' functionality for that particular buffer. If you want to rely on its functionality in any way (like described earlier), I'd suggest not use it.

Also, as this seems out of scope for 'mini.completion' features right now, I'll move this to Discussions. Feel free to continue asking question, though. Happy to help!
